npe2/_plugin_manager.py

```diff
@@ -142,7 +142,10 @@ def iter_compatible_readers(self, paths: List[str]) -> Iterator[ReaderContributi
             # but would we yield duplicate anymore.
             # above does not have have the unseen check either.
             # it's easy to make an iterable version if we wish, or use more-itertools.
-            yield from {r for pattern, r in self._readers if fnmatch(path, pattern)}
+            # match against pattern.lower() to make matching case insensitive
+            yield from {
+                r for pattern, r in self._readers if fnmatch(path, pattern.lower())
```

Comment:
```suggestion
                r for pattern, r in self._readers if fnmatch(path, pattern.lower()) or fnmatch(path, pattern.upper())
```

Reply:
huh, instead of doing .lower() on paths in iter_compatible_readers?
clever.

But what if there's mixed case?

Reply:
Maybe https://docs.python.org/3/library/fnmatch.html#fnmatch.translate and run `re.match with` https://docs.python.org/3/library/re.html#re.IGNORECASE ?

Reply:
seems really complex, is it better than just lower casing for the comparison?

Reply:
you mean `fnmatch(path.lower(), pattern.lower())`?

Reply:
yea. it's the simplest and I think it should work...

Reply:
Ok, using `path.lower(), pattern.lower()` in the fnmatch works in napari, but not in the existing tests i mimicked, because, as far as I can tell, the sample_plugin hard-codes lower case extension in the reader function:
https://github.com/napari/npe2/blob/d7329b7088f59ac743a67ff47aca32067b5a246a/tests/sample/my_plugin/__init__.py#L24-L41

My tests error with `E           ValueError: Test plugin should not receive unknown data` so I think this is actually the expected result of everything working properly.
By using `.lower()` only in the `fnmatch` the plugin gets the real path and in this case, errors. On the other hand, making the actual path have `ext.lower()` as before, would make sample_plugin handle this case and the new tests would pass.
So options are to assert something else, go back to .lower() on the path, or use .lower() in sample_plugin to make the actual plugin also case insensitive.

I think I lean towards option 1 or option 3. npe2 should be case-insensitive, but if a plugin is actually case sensitive for some reason, then it should be allowed to raise an error?

Reply:
Hmm I don't think we can/should distinguish between path being case insensitive and pattern being case insensitive. I think our reading in general should be either case sensitive or case insensitive. I think the same should be true of plugins (so, plugins should be case insensitive), but the `get_reader` functions complicate things a little, since they may be checking things about the path that are not true if we try to enforce case insensitivity by passing `path.lower()`.

I think we should make paths fully case insensitive for reading so:

- match `pattern.lower()` against `ext.lower()`
- we always pass through the unchanged path because I think that makes most sense
- add docs to contribution guide for readers to advise that pattern matching will be case insensitive
    - and say that we encourage readers to also be case insensitive
    - but if you desperately need case sensitivity you could check in your `get_reader` function (but I don't think we even mention this tbh)
- we update the cookiecutter reader to recommend case insensitivity

I had a look through all the `filename_patterns` being declared. Only 5 out of 92 readers declare any pattern with capital letters in it:

```
napari-annotatorj: ['<EDIT_ME>']
napari-deepfinder: ['*.mrc', '*.map', '*.rec', '*.h5', '*.tif', '*.TIF', '*.xml', '*.ods', '*.xls', '*.xlsx']
napari-pdr-reader: ['*.fits', '*.FITS', '*.lbl', '*.img', '*.LBL', '*.IMG']
napari-rioxarray: ['*.vrt', '*.tif', '*.tiff', '*.TIF', '*.TIFF', '*.img', '*.lbl', '*.cub', '*.fits', '*.IMG', '*.LBL', '*.CUB', '*.FITS']
napari-tomocube-data-viewer: ['*.TCF']
```

The first is clearly a placeholder. `napari-deepfinder` explicitly makes sure it's case insensitive for TIFs, `napari-pdr-reader` also makes its patterns case insensitive, and so does `napari-rioxarray` for all but one of the extensions - but it then goes through and l[owercases the path](https://github.com/AndrewAnnex/napari-rioxarray/blob/main/src/napari_rioxarray/_reader.py#L42) as it comes in anyway. I couldn't find the repo for `napari-tomocube-data-viewer` so not sure if it truly is case sensitive or not. Anyway, given the above, I would say we adopt case insensitivity for both patterns and paths, and update docs accordingly.